apollo-federation/src/sources/connect/validation/selection.rs

```diff
@@ -392,3 +394,105 @@ impl SelectionValidator<'_, '_> {
+                });
+            };
+
+            let Some(definition) = group.ty.fields.get(field_name) else {
+                return Some(Message {
+                    code: Code::SelectedFieldNotFound,
+                    message: format!(
+                        "{coordinate} contains field `{field_name}`, which does not exist on `{parent_type}`.",
+                        coordinate = selection_coordinate,
+                        parent_type = group.ty.name,
+                    ),
+                    locations: selection_location.iter().cloned().collect(),
+                });
+            };
+
+            if definition.ty.is_list() {
+                return Some(Message {
+                    code: Code::InvalidStarSelection,
+                    message: format!(
+                        "{coordinate} contains `{field_name}: *` but the field `{parent_type}.{field_name}: {ty}` returns a list. It must be a non-list custom scalar.",
+                        coordinate = selection_coordinate,
+                        field_name = field_name,
+                        parent_type = group.ty.name,
+                        ty = definition.ty
+                    ),
+                    locations: selection_location.iter().cloned().collect(),
+                });
+            }
+
+            let Some(ty) = schema.types.get(definition.ty.inner_named_type()) else {
+                return Some(Message {
+                    code: Code::GraphQLError,
+                    message: format!(
+                        "{coordinate} contains field `{field_name}`, which has undefined type `{type_name}.",
+                        coordinate = selection_coordinate,
+                        type_name = definition.ty.inner_named_type()
+                    ),
+                    locations: selection_location.iter().cloned().collect(),
+                });
+            };
+
+            if !ty.is_scalar() {
+                return Some(Message {
+                    code: Code::InvalidStarSelection,
+                    message: format!(
+                        "{coordinate} contains `{field_name}: *` but the field `{parent_type}.{field_name}: {ty}` returns {ty_kind} type. It must be a non-list custom scalar.",
+                        coordinate = selection_coordinate,
+                        field_name = field_name,
+                        parent_type = group.ty.name,
+                        ty = definition.ty,
+                        ty_kind = match ty {
+                            ExtendedType::Object(_) => "an object",
+                            ExtendedType::Interface(_) => "an interface",
+                            ExtendedType::Union(_) => "a union",
+                            ExtendedType::Enum(_) => "an enum",
+                            ExtendedType::InputObject(_) => "an input object",
+                            ExtendedType::Scalar(_) => "a scalar",
+                        }
```

Comment:
We should probably make this a helper so we can also use this in other validations. I think right now we just say "is not an object" or whatever for leaf agreement